alf/environments/suite_carla.py

```diff
@@ -1037,38 +1114,94 @@ def get_current_time_step(self, current_frame):
 
         if self._fail_frame is None:
             step_type = ds.StepType.FIRST
             max_frames = math.ceil(
-                self._current_distance / self._min_speed / self._delta_seconds)
+                self._route_length / self._min_speed / self._delta_seconds)
             self._fail_frame = current_frame + max_frames
         elif (self._current_distance < self._success_distance_thresh
               and self._actor.get_velocity() == carla.Location(0., 0., 0.)):
             # TODO: include waypoint orientation as success critiria
-            self._done = True
             step_type = ds.StepType.LAST
             reward += self._success_reward
             discount = 0.0
             info['success'] = np.float32(1.0)
+            logging.info(
+                "actor=%d frame=%d SUCCESS" % (self._actor.id, current_frame))
         elif current_frame >= self._fail_frame:
-            self._done = True
+            logging.info("actor=%d frame=%d FAILURE: out of time" %
+                         (self._actor.id, current_frame))
+            step_type = ds.StepType.LAST
+        elif (self._collision_loc is not None
+              and current_frame - self._collision_frame >
+              self._max_stuck_at_collision_frames
+              and np.linalg.norm(curr_loc - self._collision_loc) <
+              self._stuck_at_collision_distance):
+            logging.info("actor=%d frame=%d FAILURE: stuck at collision" %
+                         (self._actor.id, current_frame))
             step_type = ds.StepType.LAST
         else:
             step_type = ds.StepType.MID
 
-        prev_loc = _to_numpy_loc(self._prev_location)
-        curr_loc = _to_numpy_loc(self._actor.get_location())
-        goal0 = obs['navigation'][2]  # This is about 10m ahead in the route
-        reward += (np.linalg.norm(prev_loc - goal0) -
-                   np.linalg.norm(curr_loc - goal0))
+        if self._sparse_reward:
+            current_index = self._navigation.get_next_waypoint_index()
+            if step_type == ds.StepType.LAST and info['success'] == 1.0:
+                # Since the episode is finished, we need to incorporate the final
+                # progress towards the goal as reward to encourage stopping near the goal.
+                reward += (np.linalg.norm(self._intermediate_start -
+                                          self._goal_location) -
+                           np.linalg.norm(curr_loc - self._goal_location))
+            elif self._intermediate_goal_index < current_index:
+                # This means that the car has passed the intermediate goal.
+                # And we give it a reward which is equal to the distance it
+                # travels.
+                intermediate_goal = self._navigation.get_waypoint(
+                    self._intermediate_goal_index)
+                reward += np.linalg.norm(intermediate_goal -
+                                         self._intermediate_start)
```

Comment:
If current_index is the position on the route, then perhaps it's better to use current_pos - _intermediate_start_pos as the reward, instead of the linear distance between the two locations?

Reply:
index may not corresponds to distance if waypoints are not evenly spaced.